apis/pkg/v1beta1/image_config_types.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// MatchType is the method used to match the image.
+type MatchType string
+
+const (
+	// Prefix is used to match the prefix of the image.
+	Prefix MatchType = "Prefix"
+)
+
+// +kubebuilder:object:root=true
+// +genclient
+// +genclient:nonNamespaced
+
+// The ImageConfig resource is used to configure settings for package images.
+//
+// +kubebuilder:printcolumn:name="FIRST_PREFIX",type="string",JSONPath=".spec.matchImages[0].prefix"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories={crossplane}
+type ImageConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ImageConfigSpec `json:"spec,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ImageConfigList contains a list of ImageConfig.
+type ImageConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ImageConfig `json:"items"`
+}
+
+// ImageMatch defines a rule for matching image.
+type ImageMatch struct {
+	// Match is the method used to match the image.
+	// +optional
+	// +kubebuilder:validation:Enum=Prefix
+	// +kubebuilder:default=Prefix
+	Match MatchType `json:"match,omitempty"`
```

Comment:
nit: `spec.matchImages.match` sound weird, maybe just `spec.match.type`? image is already in the resource itself, it's an `ImageConfig` so it matches images

Reply:
or `spec.images.match`

Reply:
If we go with `spec.images.match`:

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: ImageConfig
metadata:
  name: e2e-private-provider-auth
spec:
  images:
    - prefix: xpkg.upbound.io/crossplane/e2e-private-provider
  registry:
    authentication:
      pullSecretRef:
        name: e2e-private-provider-pull-secret
```

Assuming most people will omit `match: Prefix`, relying on its default; this doesn't feel great 🤔 May be `spec.forImages.with` ?

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: ImageConfig
metadata:
  name: e2e-private-provider-auth
spec:
  forImages:
  - with: Prefix
    prefix: xpkg.upbound.io/crossplane/another
```

when omitted:

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: ImageConfig
metadata:
  name: e2e-private-provider-auth
spec:
  forImages:
  - prefix: xpkg.upbound.io/crossplane/another
```

I like it being read as `ImageConfig` `forImages` `with` `Prefix` 😄

Reply:
I don't see anything wrong with `spec.matchImages.match` personally. I'd rather stick to what we have in the design.

> Assuming most people will omit match: Prefix, relying on its default; this doesn't feel great

Generally I think it's a better practice to include `match: Prefix`. As I mentioned in the relevant design doc comment thread, this has advantages when reading the file (e.g. on disk). It's self-documenting in that you immediately see the type of match could be configurable. (Even though right now there's only the one option.)

Reply:
Fine `matchImage`, it's just repetitive being part of an ImageConfig, and I personally try to avoid multiple words in keys as camelCase for me is a source of typos... but I don't see a reason to go with `match` instead of good old `type`, it would mirror what we have for patches too. As a user I would expect to put a regex or some pattern as value for `match`. 

The design can be updated I don't see that as a valid reason.

Reply:
> I don't see a reason to go with match instead of good old type, it would mirror what we have for patches too.

This sounds good to me. I'll update `match: Prefix`  as `type: Prefix`.

Reply:
> Edit: So really I'm proposing making match a required field.

Just noticing this edit; `match` (or now `type`) is already a required field but I assume you mean not having a default for that, so, people have to write it every time instead of omitting even for prefix match. 
> this has advantages when reading the file (e.g. on disk). It's self-documenting in that you immediately see the type of match could be configurable.

Doesn't this apply to any default an API can have? 

Requiring an explicit value for a field that could have a good default just to make it visible in every possible case and self documenting feels like an unnecessary overhead to me. If we end up never introducing another type, or using `Prefix` in 95% of the cases, it is even worse.

Reply:
> Doesn't this apply to any default an API can have?

No. If there's a required field with a default, I can write this on in my git repo and then apply it:

```yaml
matchImages:
- prefix: foo
```

A reader would never know there was a type field with other options for the type. The document is less self documenting, and becomes asymmetrical if/when other types are needed. (Sometimes you need to specify a type, sometimes you don't.)

This isn't a hill I'll die on, but I have a hunch it's something we'll regret in time _if_ we end up adding other types.